### PR TITLE
Added support for the 'Link' header for content negotiation.

### PR DESCRIPTION
**Pull request by tomgillett**
schema.org recently dropped support for content negotiation. They instead adopted a Link header, with `rel="alternate"` and `type="application/ld+json"`. This attempts to utilise that header.

See: https://github.com/schemaorg/schemaorg/issues/2578#issuecomment-632227864

### CODE REVIEW

File: FileGetContentsLoader.php [hunk 1]
```diff
@@ -27,95 +27,134 @@ public function loadDocument($url)
...
+                $remoteDocument->mediaType = substr($remoteDocument->mediaType, 0, $pos);
             }
 
-            $remoteDocument->document = Processor::parse($input);
+            $remoteDocument->mediaType = trim($remoteDocument->mediaType);
 
-            return $remoteDocument;
+            if ('application/ld+json' === $remoteDocument->mediaType) {
+                $remoteDocument->contextUrl = null;
+            } elseif ('application/octet-stream' === $remoteDocument->mediaType) {
```
Comment (tomgillett) [inline FileGetContentsLoader.php:143]: This is not ideal, but appears to be what schema.org actually returns...

Comment (Sarke) [reply FileGetContentsLoader.php:143]: This was fixed:

https://github.com/schemaorg/schemaorg/issues/2595

Comment (shawnhind) [reply FileGetContentsLoader.php:143]: I still get the error saying it is getting text/html back.

Comment (shawnhind) [reply FileGetContentsLoader.php]: Oh I see maybe you were saying that specific part of the problem was fixed.

**Comment by tomgillett** [comment]
@Sarke thanks for letting me know about https://github.com/schemaorg/schemaorg/issues/2595. I've tested and schema.org is now returning the correct content type (it wasn't yesterday). I've removed `application/octet-stream` accordingly, which makes me much happier!

**Comment by rvanlaak** [comment]
This repo's latest commit is of November 2018, so I took the liberty to Tweet @lanthaler to try getting some traction here to get related fixes merged and released: https://twitter.com/Rvanlaak/status/1268460128963084288

File: FileGetContentsLoader.php [hunk 2]
```diff
@@ -27,95 +27,132 @@ public function loadDocument($url)
...
-                'http' => $streamContextOptions,
-                'https' => $streamContextOptions
-            ));
+        // Extract HTTP Link headers
+        $linkHeaderValues = array();
+        if (is_array($http_response_header)) {
+            for ($i = count($http_response_header) - 1; $i > $httpHeadersOffset; $i--) {
+                if (0 === substr_compare($http_response_header[$i], 'Link:', 0, 5, true)) {
+                    $value = substr($http_response_header[$i], 5);
+                    $linkHeaderValues = array_merge($linkHeaderValues, explode(',', $value));
```
Comment (lanthaler) [inline FileGetContentsLoader.php:83]: Why is this change required?

File: FileGetContentsLoader.php [hunk 3]
```diff
@@ -27,95 +27,132 @@ public function loadDocument($url)
...
-                    if (0 === substr_compare($http_response_header[$i], 'Link:', 0, 5, true)) {
-                        $value = substr($http_response_header[$i], 5);
-                        $linkHeaderValues[] = $value;
+                if (isset($link['rel']) && isset($link['type']) && ($link['rel'] === 'alternate')) {
+                    switch($link['type']) {
+                        case 'application/ld+json':
+                        case 'application/json':
+                            return $this->retrieveRemoteDocument($link['uri']);
+                            break;
                     }
```
Comment (lanthaler) [inline FileGetContentsLoader.php:115]: Please generalize parseContextLinkHeaders() instead of duplicating most of its code and then please move the handling of the alternate link header to down to where we check the media type. This should only trigger if we received something other than a JSON-LD doc

Comment (tomgillett) [reply FileGetContentsLoader.php:115]: @lanthaler I'm very happy to take a look at this, but do you have any example headers I can use as text fixtures in this refactor? I avoided eding `parseContextLinkHeaders()` as I did not want to accidentally break existing functionality.

**Comment by tomgillett** [comment]
> If you have a little bit of extra time, it would be fantastic if you could also add a unit test for the Link header parsing.

@lanthaler I'm happy to do this. 

Is there a reason why `parseContextLinkHeaders()` is private-scope? I'll need to change this to public in order to unit test.

A cleaner approach would be to leave it private (or protected?) and refactor the class to use an HTTP client (rather than `file_get_contents`) so we can use a mock client with response... but I think this would be a much bigger refactor. I can see this was originally considered in https://github.com/lanthaler/JsonLD/issues/4. [PSR-17](https://www.php-fig.org/psr/psr-17/) and [PSR-18](https://www.php-fig.org/psr/psr-18/) might also be helpful for this.

File: FileGetContentsLoader.php [hunk 4]
```diff
@@ -119,44 +134,40 @@ function ($code, $severity, $msg, $msgCode, $bytesTx, $bytesMax) use (
...
+
+                    $links[] = $link;
+                }
             }
         }
 
-        return array_values(array_unique($contexts));
+        return $links;
     }
 }
```
Comment (lanthaler) [inline FileGetContentsLoader.php:173]: Hmm... do you really need to change all of this. AFAICT, the only change necessary should be to change how the URI is handled in `$part` below and return all parts

```php
        // Split the header in key-value pairs
        foreach ($values as $val) {
            $part = array();
            foreach (preg_split('/;(?=([^"]*"[^"]*")*[^"]*$)/', $val) as $kvp) {
                preg_match_all('/<[^>]+>|[^=]+/', $kvp, $matches);
                $pieces = array_map($trimWhitespaceCallback, $matches[0]);
                $part[$pieces[0]] = isset($pieces[1]) ? $pieces[1] : '';
            }

            if (in_array('http://www.w3.org/ns/json-ld#context', explode(' ', $part['rel']))) {
                $contexts[] = (string) $baseIri->resolve(trim(key($part), '<> '));
            }
        }

        return array_values(array_unique($contexts));
```

Something like this

```php
        $result = array();

        // Split the header in key-value pairs
        foreach ($values as $val) {
            $part = array();
            foreach (preg_split('/;(?=([^"]*"[^"]*")*[^"]*$)/', $val) as $kvp) {
                preg_match_all('/<[^>]+>|[^=]+/', $kvp, $matches);
                $pieces = array_map($trimWhitespaceCallback, $matches[0]);
->              if (!isset($pieces[1])) {
->                 $part['uri'] = (string) $baseIri->resolve(trim($pieces[0], '<> '));
->              } else {
->                  $part[$pieces[0]] = $pieces[1];
->              }
            }
->          $result[] = $part;
        }

->      return $result;
```

File: FileGetContentsLoader.php (hunk 3 above)
Comment (lanthaler) [reply FileGetContentsLoader.php]: Thanks. I'd suggest to reduce the number of changes to a minimum (see other comment) as this code has proven quite stable over the years and already does all the parsing you need.

File: Test/FileGetContentsLoaderTest.php [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+    protected $iri;
+
+    protected $loader;
+
+    public function setUp()
+    {
+        parent::setUp();
+
+        $this->iri = new IRI('https://www.google.com');
```
Comment (lanthaler) [inline Test/FileGetContentsLoaderTest.php:29]: Even though I work at Google these days I'd prefer to use example.com here and elsewhere :-)

File: FileGetContentsLoader.php (hunk 4 above)
Comment (tomgillett) [reply FileGetContentsLoader.php]: I've adopted this approach in latest set of commits (mostly as you suggested, but tweaked a little as my unit tests revealed a few issues 👍)

File: Test/FileGetContentsLoaderTest.php (hunk 5 above)
Comment (tomgillett) [reply Test/FileGetContentsLoaderTest.php]: Heh! I'd not considered that. I've switched to foobar.com and fizzbuzz.net 😆

File: FileGetContentsLoader.php [hunk 6]
```diff
@@ -100,13 +104,24 @@ function ($code, $severity, $msg, $msgCode, $bytesTx, $bytesMax) use (
...
+                } else {
+                    // If the Media type was not as expected, check to see if the desired content type
+                    // is being offered in a Link header (this is what schema.org now does).
+                    $altLinkHeaders = array_filter($linkHeaderValues, function ($link) {
+                        return (isset($link['rel']) && isset($link['type']) 
+                            && ($link['rel'] === 'alternate') && ($link['type'] === 'application/ld+json'));
+                    });
+
+                    if (count($altLinkHeaders) && $altLinkHeaders[0]['uri']) {
+                        return $this->loadDocument($altLinkHeaders[0]['uri']);
```
[the remark this answers is not in the document]
Comment (tomgillett) [reply FileGetContentsLoader.php:118]: Done! 👍

**Comment by tomgillett** [comment]
@lanthaler I think I've resolved that latest round of feedback. Sticking with your original parsing logic is the safer approach and so I'm happier where we've landed!

> Absolutely. I started moving this into https://github.com/lanthaler/fgc-client a few years ago before those PSRs existed and used HTTPlug instead. If you are interested, we should update that package and then have JsonLD have depend on it and allow people to use other clients. The reason I like file_get_contents is that it supports both local and remote files and is available basically everywhere incl. the cheapest shared hosts.

I'll pick this thread up over on https://github.com/lanthaler/JsonLD/issues/4 😊
